internal/rpmmd/repository.go

```diff
@@ -534,6 +543,155 @@ func (r *rpmmdImpl) Depsolve(packageSet PackageSet, repos []RepoConfig, modulePl
+// on the order of package set names passed in `packageSetsChain`. Repositories
+// provided in `repos` are used for every transaction, while repositories from
+// `packageSetsRepos` are used only for package sets with the respective name.
+//
+// The function returns a `ChainPackageSet` slice, which members are referencing
+// repositories from the returned `RepoConfig` by their index. The returned
+// `RepoConfig` slice is specific to the requested `packageSetsChain`.
+//
+// NOTE: Due to implementation limitations of DNF and dnf-json, each package set
+// in the chain must use all of the repositories used by its predecessor.
+// An error is returned if this requirement is not met.
+func chainPackageSets(packageSetsChain []string, packageSets map[string]PackageSet, repos []RepoConfig, packageSetsRepos map[string][]RepoConfig) ([]chainPackageSet, []RepoConfig, error) {
+	transactions := make([]chainPackageSet, 0, len(packageSetsChain))
+	transactionsRepos := make([]RepoConfig, 0, len(repos))
+
+	transactionsRepos = append(transactionsRepos, repos...)
+
+	// These repo IDs will be used for every transaction
+	baseRepoIDs := make([]int, 0, len(repos))
+	for idx := range repos {
+		baseRepoIDs = append(baseRepoIDs, idx)
+	}
+
+	for transactionIdx, pkgSetName := range packageSetsChain {
+		pkgSet, ok := packageSets[pkgSetName]
+		if !ok {
+			return nil, nil, fmt.Errorf("package set %q requested in the 'packageSetsChain' does not exist in provided 'packageSets'", pkgSetName)
+		}
+
+		transaction := chainPackageSet{
+			PackageSet: pkgSet,
+			Repos:      baseRepoIDs, // Due to its capacity, the slice will be copied if any repo is appended
+		}
+
+		// Add any package-set-specific repos to the list of transaction repos
+		if pkgSetRepos, ok := packageSetsRepos[pkgSetName]; ok {
+			for _, pkgSetRepo := range pkgSetRepos {
+				// Check if the repo has been already used by a transaction
+				// and if yes, just use its ID. Skip the "base" repos.
+				pkgSetRepoID := -1
+				for idx := len(repos); idx < len(transactionsRepos); idx++ {
+					transactionRepo := transactionsRepos[idx]
+					if reflect.DeepEqual(pkgSetRepo, transactionRepo) {
+						pkgSetRepoID = idx
+						break
+					}
+				}
+
+				if pkgSetRepoID == -1 {
+					transactionsRepos = append(transactionsRepos, pkgSetRepo)
+					pkgSetRepoID = len(transactionsRepos) - 1
+				}
+
+				transaction.Repos = append(transaction.Repos, pkgSetRepoID)
+			}
+		}
+
+		// Sort the slice of repo IDs to make it easier to compare
+		sort.Ints(transaction.Repos)
+
+		// If more than one transaction, ensure that the transaction uses
+		// all of the repos from its predecessor
+		if transactionIdx > 0 {
+			previousTransRepos := transactions[transactionIdx-1].Repos
+			if len(transaction.Repos) < len(previousTransRepos) {
+				return nil, nil, fmt.Errorf("chained packageSet %q does not use all of the repos used by its predecessor", pkgSetName)
+			}
+
+			for idx, repoID := range previousTransRepos {
```

Comment:
This is basically an equality test for two subslices, might be cleaner to write it in that way (another `reflect.Deep`)

Reply:
Yes, it is an equality check of a slice (previous transaction) and subslice (current transaction capped at the length of the previous transaction).

Happy to change it if you feel like using more of `reflect` 😄 .

I didn't use it in this case as implementing this check was considerably simpler than deep equality of two structures.

Reply:
It just might be simpler to read but no hard feelings. :)